File: src/org/grumblesmurf/df/careers/DwarfRenderer.java
package org.grumblesmurf.df.careers;

import javax.swing.*;
import java.awt.*;

public class DwarfRenderer implements ListCellRenderer
{
    private final ImageIcon male = createImageIcon("/icons/user.png", "Male");
    private final ImageIcon female = createImageIcon("/icons/user_female.png", "Female");

    @Override
    public Component getListCellRendererComponent(JList list, Object value, int index, boolean isSelected,
                                                  boolean cellHasFocus) {
        Dwarf dwarf = (Dwarf) value;
        JLabel jLabel = new JLabel(value.toString());
        jLabel.setIcon(createImageIconFor(dwarf));
        if (isSelected) {
            jLabel.setOpaque(true);
            jLabel.setBackground(list.getSelectionBackground());
            jLabel.setForeground(list.getSelectionForeground());
        } else {
            jLabel.setBackground(list.getBackground());
            jLabel.setForeground(list.getForeground());
        }
        return jLabel;
    }

    private Icon createImageIconFor(Dwarf dwarf) {
        return "Female".equals(dwarf.sex) ? female : male;
    }

    protected ImageIcon createImageIcon(String path, String description) {
        java.net.URL imgURL = getClass().getResource(path);
        return new ImageIcon(imgURL, description);
    }
}

File: src/org/grumblesmurf/df/careers/DwarfRenderer.scala
package org.grumblesmurf.df.careers

import javax.swing._
import java.awt._

class DwarfRenderer extends ListCellRenderer {
  def getListCellRendererComponent(list: JList, value: AnyRef, index: Int, isSelected: Boolean, cellHasFocus: Boolean): Component = {
    val dwarf = value.asInstanceOf[Dwarf]
    val jLabel = new JLabel(value.toString)
    jLabel.setIcon(createImageIconFor(dwarf))
    if (isSelected) {
      jLabel.setOpaque(true)
      jLabel.setBackground(list.getSelectionBackground)
      jLabel.setForeground(list.getSelectionForeground)
    } else {
      jLabel.setBackground(list.getBackground)
      jLabel.setForeground(list.getForeground)
    }
    jLabel
  }

  private def createImageIconFor(dwarf: Dwarf) = {
    if (("Female" == dwarf.sex)) female else male
  }

  protected def createImageIcon(path: String, description: String) = {
    val imgURL = getClass.getResource(path)
    new ImageIcon(imgURL, description)
  }

  private final val male = createImageIcon("/icons/user.png", "Male")
  private final val female = createImageIcon("/icons/user_female.png", "Female")
}
